Fix error recovery after errors during `SSTextField` edit

errael opened the pull request:
Fix #175
Return boolean from `updateColumnText` indicating if an error occurred.
Add `DocumentFilter` to capture text field value before edit.
Use `SSSQLNullException` instead of `NPE`.
Deprecate methods that change `SSCommon` or `ssComponent`.
Add some constraints for `supplier_data` table to make it easier to test error recovery.

errael committed:
Make status "not null" in database.

bpangburn commented on swingset/src/main/java/com/nqadmin/swingset/SSTextField.java:
```diff
@@ -478,10 +494,13 @@ public void setNumberOfDecimalPlaces(final int _numberOfDecimalPlaces) {
 	 * @param _ssCommon shared/common SwingSet component data and methods
+	 * @deprecated Creates inconsistent state, no replacement.
 	 */
+	// TODO: is this used? What about all the listeners and such?
```
I think SSComponentInterface() could be modified to remove setSSCommon(). I did a reference search in Eclipse and its not used from within SwingSet. I also don't see references in our software using SwingSet.

I don't see any reason ssCommon would need to be replaced once a SwingSet component is instantiated.

Best guess here is that when I created SSCommon and SSComponentInterface, I was creating setter/getter pairs to be thorough, but I agree that it's dangerous to call setSSCommon().

errael commented:
Split and rebased. The only change I want to make is disabling events in `SSCommon::changedUpdate()`
```
if (!ok) {
        // restore previous text value
        if(previousValue != null) {
                // TODO: should disable document events similar
                // to how row set events are handled.
                // can make that change if needed/requested/required.
                ((JTextComponent) getSSComponent()).setText(previousValue);
        }
}
```
And clean up some leftover TODOs...

bpangburn commented on swingset/src/main/java/com/nqadmin/swingset/SSTextField.java:
```diff
@@ -287,21 +296,44 @@ public SSTextField(final int _mask, final int _numberOfDecimalPlaces, final int
+	/** All the contructors feed through here */
+	private SSTextField(String _text, final RowSet _rowSet, final String _boundColumnName,
+	final int _mask, final int _numberOfDecimalPlaces, final int _align) {
 		super(_text);
```
I wonder if we need both ssCommon and constructingSSCommon.

Rather than this:
```
	private SSCommon constructingSSCommon() {
		if (constructingSSCommon == null) {
			constructingSSCommon = new SSCommon(this);
		}
		return constructingSSCommon;
	}
```

Could we use this:
```
	private SSCommon constructingSSCommon() {
                SSCommon result = ssCommon;

		if (ssCommon == null) {
			result = new SSCommon(this);
		}
		return result;
	}
```

errael replied:
`constructingSSCommon()` is used from `createDefaultModel()` which needs to use the same SSCommon that ends up in the variable ssCommon. 

The real issue is that `JTextField` calls `createDefaultModel()` in its constructor; and the `SSTextField` constructor does `super(_text)` so the document model, which depends on sscommon, has to be in place before `SSTextField` is constructed. 

An alternative approach would be to let `JTextField` use it's own document model, then in `SSTextField`'s constructor create the real model, based on sscommon, copy the text from the `JTextField`'s model into the real model, and set the new model into `SSTextField`. That would be more work (creating two models) and since I tend to overoptimize...

I added a comment, which _may_ clarify things at least a little bit.

bpangburn commented on swingset/src/main/java/com/nqadmin/swingset/SSTextField.java:
```diff
@@ -287,21 +296,44 @@ public SSTextField(final int _mask, final int _numberOfDecimalPlaces, final int
-	 * @param _mask the mask required for this textfield.
-	 */
-	public SSTextField(final String _text, final int _mask) {
+	/** All the contructors feed through here */
```
All the **constructors**...(misspelling)

bpangburn commented on swingset/src/main/java/com/nqadmin/swingset/utils/SSCommon.java:
```diff
@@ -157,9 +166,29 @@ public void changedUpdate(final DocumentEvent de) {
+							// restore previous text value
+							if(previousValue != null) {
+								if (ssComponentListenerAdded) {
+									// avoid generating events while restoring text
```
Should we manage `listenerNeedsRestoration` inline here or should we move a state tracking variable to addSSDocumentListener() and removeSSDocumentListener() similar to addSSComponentListener() and removeSSComponentListener()?

This way you can call addSSDocumentListener() and it will add the missing listener if listenerNeedsRestoration is true and then set it to false once complete.

errael replied:
I don't know if that works; I'm not saying it doesn't. I'm not sure exactly what you are suggesting.

Is it possible that `addSSDocumentListener()` would be called while `listenerNeedsRestoration` is true? If it is, then I'm just confusing myself. But I don't see how that could happen.

I'm still not sure, in detail, what you're suggesting. This is more your bailiwick, LMK.

errael replied:
Added a commit, https://github.com/bpangburn/swingset/pull/178/commits/eac0fc1dc82a2644121f0aa4f0238449138b4ded, with a TODO for this.

bpangburn reviewed (commented):
Made a comment where we _might_ be able to eliminate a variable.

Found one misspelling around line 303 of SSTextField.

Tested a bit against PostgreSQL and seems to be working fine.

errael committed:
Add/correct comments after review.

errael commented:
Added https://github.com/bpangburn/swingset/pull/178/commits/eac0fc1dc82a2644121f0aa4f0238449138b4ded to address review comments.

bpangburn reviewed (approved):
Approved pending @errael thoughts on extending this fix for SSTextArea and SSFormattedTextField.
